docs/CIP/CIP-XXXX/README.md

```diff
@@ -0,0 +1,272 @@
+
+### Versioning
+
+In our protocol, the client sends its list of supported versions as part of the initial [`subscribe`](./messages/client/subscribe.md) message. Upon receiving this information, the server selects the highest mutually supported version and associates it with the client.
+This method requires the server to implement version-based message construction, allowing it to tailor its responses to each client's supported version.
+
+### Encoding
+
+Our proposed communication protocol **adopts blockchain-native encoding standards** to ensure compatibility and efficiency. This approach leverages existing conventions familiar to clients while minimizing conversion overhead on the server-side.
+
+#### Transaction Hash Encoding
+
+Transaction hashes are typically encoded using **hexadecimal representation**.
+
+#### Address Encoding
+
+Address encoding varies depending on the blockchain network. In Cardano the protocol servers bech32 encoded addresses.
+
+### Extensions
+
+Clients can turn on certain `extensions` by setting a respective flags in their subscription [`config](./messages/client/subscribe.md#subscribe-example) which tells the server to for example:
+
+- to enrich transaction output asset values by their respective off-chain metadata such as decimal offsets
+- to resolve transaction inputs (transaction output references) as transaction outputs
+
+## Rationale: how does this CIP achieve its goals?
+
+A primary consideration for the protocol design was to minimize message round trips between client and server as well as the amount of work to be done by the server for each client. Thus, the API's authentication does not require an initial message exchange but rather defines a standard verifable method for the server to either drop a client early and proceed to publish its relevant data.
+
+### Client Typed Messages
+
+- [`heartbeat`](./messages/client/heartbeat.md): Connection management/ keep alive
+- [`subscribe`](./messages/client/subscribe.md): Define topics of interest & authentication
+- [`unsubscribe`](./messages/client/unsubscribe.md): Graceful disconnect from server
+
+### Server Message Partials
+
+- [`genesis`](./messages/server/genesis.md): If applicable for the subscribed blockchain, it serves static data used as part of the network bootstrap
+- [`transaction`](./messages/server/transaction.md): A new transaction
+- [`tip`](./messages/server/tip.md): A new block was appended
+- [`protocol parameters`](./messages/server/protocol-parameters.md): Epoch based message for updated protocol parameters
+- [`era summary`](./messages/server/era-summary.md): Era based message for updated slot length
```

Comment:
Since the credentials are no longer supplied by the client, server should send the credentials it discovered. Otherwise client won't know it's own addresses.
```suggestion
- [`era summary`](./messages/server/era-summary.md): Era based message for updated slot length
- credentials: credentials discovered by the server
```

Reply:
Sorry @mkazlauskas , I do not understand this.

> Since the credentials are no longer supplied by the client

Didn't we say that clients subscribe through credentials?

I thought the flow should be something similar to:

1. the client asks the server to follow all transactions relevant for all its credentials;
  - `subscribe` with a set of credentials (so credentials are provided by the client);
2. the server answers with the list of all addresses bound to the credentials set;
  - having transactions indexed by credentials, the server can find all of them using the indexes rather than the standard discovery process;
  - rather than adding `credentials: credentials discovered by the server` probably we can add `addresses: addresses discovered by the server`.

Reply:
> Didn't we say that clients subscribe through credentials?

That was the original idea, but then we added subscribing through xpub which is not a credential, but is a base key that can be used to derive the key hash credentials